src/content/ui/adaptive-responsive/layout.md

```diff
@@ -0,0 +1,494 @@
+
+The [Folio source code][] is also available on GitHub.
+Learn more on the [gskinner blog][].
+
+[Adaptive layouts]: {{site.yt.watch}}?v=n6Awpg1MO6M&t=694s
+[Adaptive layouts, part 2]: {{site.yt.watch}}?v=eikOZzfc0l4&t=11s
+[Folio source code]: {{site.github}}/gskinnerTeam/flutter-folio
+[gskinner blog]: https://blog.gskinner.com/
+
+### Other resources
+
+You can learn more about creating platform-adaptive apps
+in the following resources:
+
+* [Platform-specific behaviors and adaptations][],
+  a page on this site.
+* [Designing truly adaptive user interfaces][],
+  presented at the Flutter Vikings 2020 conference
+  by Aloïs Deniel,
+
+[Designing truly adaptive user interfaces]: https://www.aloisdeniel.com/blog/designing-truly-adaptative-user-interfaces
+[Platform-specific behaviors and adaptations]: /platform-integration/platform-adaptations
+
+### Contextual layout
+
+If you need more than density changes and can't find a
+widget that does what you need, you can take a more
+procedural approach to adjust parameters, calculate sizes,
+swap widgets, or completely restructure your UI to suit
+a particular form factor.
+
+#### Use LayoutBuilder for extra flexibility
+
+Even though checking total screen size is great for
+full-screen pages or making global layout decisions,
+it's often not ideal for nested subviews.
+Often, subviews have their own internal breakpoints
+and care only about the space that they have available to render.
+
+The simplest way to handle this in Flutter is using the
+[`LayoutBuilder`][] class. `LayoutBuilder` allows a
+widget to respond to incoming local size constraints,
+which can make the widget more versatile than if it
+depended on a global value.
+
+The previous example could be rewritten using `LayoutBuilder`:
+
+<?code-excerpt "lib/widgets/extra_widget_excerpts.dart (LayoutBuilder)"?>
+```dart
+Widget foo = LayoutBuilder(builder: (context, constraints) {
+  bool useVerticalLayout = constraints.maxWidth < 400;
+  return Flex(
+    direction: useVerticalLayout ? Axis.vertical : Axis.horizontal,
+    children: const [
+      Text('Hello'),
+      Text('World'),
+    ],
+  );
+});
+```
+
+This widget can now be composed within a side panel,
+dialog, or even a full-screen view,
+and adapt its layout to whatever space is provided.
+
+#### Device segmentation
+
+There are times when you want to make layout decisions
+based on the actual platform you're running on,
+regardless of size. For example, when building a
+custom title bar, you might need to check the operating
+system type and tweak the layout of your title bar,
+so it doesn't get covered by the native window buttons.
+
+To determine which combination of platforms you're on,
+you can use the [`Platform`][] API along with the `kIsWeb` value:
+
+<?code-excerpt "lib/global/device_type.dart (Platforms)"?>
+```dart
+bool get isMobileDevice => !kIsWeb && (Platform.isIOS || Platform.isAndroid);
```

Comment:
We now know this is an anti pattern. Unless you have strong feelings about it  I think tyler or myself should rewrite the "Device Segmentation" section. 

Tyler I think this is where we talk about capabilities vs policy. 

A specific exampel that came up in @khanhnwin's talk was about when to use pull to refresh. 
Anti pattern 1 use Platform.isIOS || Platfrom.isAndroid directly in code
Anti pattern 2 use isMobileDevice like is listed here. 

Recommended pattern create a policy.dart 
```
bool get shouldUsePullToRefresh() { 
    if (kIsWeb) {
        // Web has a different expectation for how to refresh content. 
        return false; 
    }
    if (Platform.isIOS ) {
        // Assume all ios devices have a touch screen. 
        return true; 
    }
    if(Platfrom.isAndroid) {
        // TODO rely on plugin to call getPackageManager().hasSystemFeature("android.hardware.touchscreen");
        // When android tv support is added
       return true;
    }
    // fallback for desktop platforms 
    // TODO when desktop is supported in our app use plugin to decide if device has a touchscreen. 
    return false; 
}
```

Reply:
I think linking out to your talk from last year and giving a full capabilities class and policy class here would be good. And add a link to your talk at the end like the other sections do

Reply:
@sfshaza2 here what should probably be its own page 

## Design to the strengths of each device type

Consider the unique strengths and weaknesses of different devices. Beyond their screen size and inputs (e.g., touch, mouse, keyboard), what other unique capabilities can you leverage?. Flutter enables our  code to *run* on different devices but strong design is more than just running code. Think about what each platform does best and see if there are unique capabilities to leverage.

For example: Apple's App Store and Google’s Play Store have different rules that apps need to abide by. Different host operating systems have differing capabilities across time as well as each other. 

Another example is leveraging the web's extremely low barrier for sharing. If we're deploying a web app, decide which deep links to support, and design the navigation routes with those in mind.

### Capabilities and Policies

Flutter's recommended pattern for handling different behavior based on these unique capabilities is to create a set of Capabilities and Policies for our app.

#### Capabilities


A Capability defines what the code or device *can* do. Examples of capabilities are, the existence of APIs, OS enforced restrictions, physical hardware requirements (like a camera). 

#### Policies

A Policy defines what the code *should* do. Examples of policies include, App store guidelines, design preferences, asset or copy that need to refer to the host device, features enabled server side. 

### How to structure the forked device code. 

The simplest mechanical way is `Platform.isAndroid` `Platfrom.isIOS` and `kIsWeb`. These APIs mechanically let us know where the code is running but have some problems as the app expands where it is run and as host platforms add functionality. 

#### Avoid using `Platform.isAndoid` and similar functions to make layout decisions or assumptions about what a device can do. 

Instead, describe what we want to branch on in a method. 

Example: Our app has a link to buy something on a website but we do not wish to show that link on ios devices for policy reasons. 
```dart
bool shouldAllowPurchaseClick() {
    // Banned by apple app store guidelines. 
    return !Platform.isIOS
}
...
TextSpan(
                  text: 'Buy in browser',
                  style: new TextStyle(color: Colors.blue),
                  recognizer: shouldAllowPurchaseClick ? TapGestureRecognizer()
                    ..onTap = () { launch('<some url>') : null;
                  } : null,

```

What did we get by adding one additional layer of indirection? The code makes it more clear why the branched path exists. This method can be in the class directly but it is likely that other areas of code might need this same check. If so, we put the code in a class. 

```dart 
/// policy.dart

class Policy {

  bool shouldAllowPurchaseClick() {
    // Banned by Apple App Store guidelines. 
    return !Platform.isIOS
  }

}
```

With this code in a class any widget test we write can now mock `Policy().shouldAllowPurchaseClick` and verify the behavior independently of the device the test is run on. 
It also means that later if we decided that buying on the web wasn't the right flow for Android users we could change the implementation and the tests for clickable text would not need to change. 


## Capabilities 
Sometimes we want our code to do something but the apis do not exist or maybe we are depending on a plugin feature that is not yet implemented on one of the platforms we support. This is a limitation of what the device *can* do. 

Those situations are similar to the policy decisions described above. We refer to those as Capabilities. Why separate them when the structure of the classes is similar? Flutter has found with productionized apps that making a logical distinction between what apps *can* do and what they *should* do helps larger products respond to changes in what platforms can do or require in addition to your own preferences after the initial code was written. 

Let's look at the case where one platform adds a new permission that requires users to interact with a system dialog before our code calls some sensitive api. Our team does the work for Platform1 and creates a Capability named `requirePremissionDialogFlow`. Then if/when platform 2 adds a similar requirement but only for new api versions then our implementation of `requirePremissionDialogFlow` can now check the API level and return true for platform 2. We already did the work, why make it hard when we need to do the same thing again. 

## Policies 

We encourage starting with a Policy class initially even if it seems like you wont make many policy based decisions. As the complexity of the class grows or the number of inputs expands we may decide to break up the policy class by feature or some other criteria.  

For policy implementation you can use compile time, run time and/or RPC backed.
Compile time policy checks are good for platforms where the preference is unlikely to change and accidentally changing the value might have large consequences. For example if a platform required that you not link to the play store or required that you use a specific payment provider given the content of your app.
Runtime checks can be good for determining if there is a touch screen the user can use. Android has a feature you can check and your web implementation could check for max touch points. 
Rpc backed policy changes are good for incremental feature rollout or for decisions that might change later. 


## Summary 

Use Capability class to define what the code *can* do. Checked against the existence of APIs, OS enforced restrictions, and physical hardware requirements (like a camera). Usually compile or runtime checks.

Use a Policy class (or classes depending on complexity) to define what the code *should* do to follow App store guidelines, design preferences, and assets or copy that need to refer to the host device. Can be a mix of compile, runtime or RPC checks. 

Test the branching code by mocking Capabilities/Policies so the widget tests do not need to change when the capabilities or policies change.

Name methods in Capabilities/Policies based not on device type but what they are trying to branch.